File: sdk/conformance/typescript/src/streaming-uploads.test.ts
import * as assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";
import { LoonFSClient } from "../../../generated/typescript/transfers.js";
import { LoonFSClient as BrowserClient } from "../../../generated/typescript-client/transfers.js";
import { UploadSource, TransferScope } from "../../../generated/typescript/transfer-runtime.js";

type Fixture = {
    name: string;
    content: string;
    algorithm: string;
    checksum: string;
    size_bytes: number;
    size: number | null;
    mode: string;
    fault?: string;
    error: boolean;
};
const cases: Fixture[] = JSON.parse(
    readFileSync(join(__dirname, "../../../../../fixtures/streaming_uploads.json"), "utf8"),
);
for (const browser of [false, true])
    for (const fixture of cases) {
        test(`streaming upload ${browser ? "browser" : "server"} ${fixture.name}`, async () => {
            let position = 0,
                eof = false,
                closed = false;
            const content = new TextEncoder().encode(fixture.content);
            const source = {
                async *[Symbol.asyncIterator]() {
                    try {
                        while (position < content.length) {
                            const start = position;
                            position += Math.min(2, content.length - position);
                            yield content.subarray(start, position);
                        }
                        eof = true;
                        if (fixture.fault === "source_error")
                            throw new Error("source failed after its last byte");
                    } finally {
                        closed = true;
                    }
                },
            };
            const counts = { payload: 0, abort: 0, complete: 0 };
            const bodies: Uint8Array[] = [];
            const claim = {
                kind: "blob",
                content_id: "cnt_00000000000000000000000000000001",
                size_bytes: fixture.size_bytes,
                checksum: { algorithm: fixture.algorithm, value: fixture.checksum },
            };
            const session = {
                namespace_id: "demo",
                namespace_alias: "demo",
                upload_id: "upl_test",
                mode: fixture.mode,
            };
            const access = {
                kind: "presigned_url",
                method: "PUT",
                url: "http://objects.test/object",
                expires_at_ms: 2000000000000,
            };
            const send: typeof fetch = async (input, init) => {
                const url = new URL(input instanceof Request ? input.url : input.toString());
                const path = url.pathname,
                    headers = new Headers(init?.headers);
                assert.ok(init?.signal, "every transport needs a signal");
                if (path === "/object") {
                    assert.equal(headers.has("authorization"), false);
                    assert.equal(headers.has("X-Private"), false);
                } else {
                    if (!browser) assert.equal(headers.get("authorization"), "Bearer private-token");
                    if (!path.endsWith("/capabilities"))
                        assert.ok(
                            path.startsWith(browser ? "/v0/namespace-aliases/demo/" : "/v0/namespaces/demo/"),
                        );
                }
                // Construct a real Request: Node rejects a stream if duplex was lost.
                const request = new Request(url, init);
                let value: unknown;
                if (path.endsWith("/capabilities"))
                    value = {
                        protocol_version: "v0",
                        api_groups: ["filesystem/v0"],
                        features: {
                            "filesystem.uploads.direct_put": fixture.mode === "direct_put",
                            "filesystem.uploads.direct_multipart": fixture.mode === "direct_multipart",
                        },
                        limits: fixture.mode === "direct_put" ? { "upload.max_content_bytes": 0 } : {},
                    };
                else if (path.endsWith("/uploads")) {
                    assert.equal((await request.json()).mode, fixture.mode);
                    value = { ...session, checksum_algorithm: fixture.algorithm, part_size_bytes: 4, access };
                } else if (path.endsWith("/parts")) {
                    const { parts } = await request.json();
                    assert.equal(parts.length, 1);
                    assert.ok(position <= bodies.length * 4 + 4, "send a part before reading the next");
                    value = { ...session, parts: [{ part_number: parts[0].part_number, access }] };
                } else if (path === "/object" || path.endsWith("/content")) {
                    counts.payload++;
                    if (fixture.size !== null)
                        assert.ok(
                            !(init?.body instanceof ReadableStream),
                            "small known bodies must remain portable in browsers",
                        );
                    bodies.push(new Uint8Array(await request.arrayBuffer()));
                    if (fixture.fault === "timeout")
                        return new Promise<Response>((_, reject) => {
                            if (init!.signal!.aborted) reject(init!.signal!.reason);
                            else
                                init!.signal!.addEventListener("abort", () => reject(init!.signal!.reason), {
                                    once: true,
                                });
                        });
                    if (fixture.fault === "payload_error") return new Response(null, { status: 503 });
                    return Response.json(
                        { ...session, content_ref: claim },
                        { headers: { ETag: "test-etag" } },
                    );
                } else if (path.endsWith("/abort")) {
                    counts.abort++;
                    value = { ...session, status: "aborted", aborted_at_ms: 0 };
                } else if (path.endsWith("/complete")) {
                    counts.complete++;
                    assert.ok(eof);
                    const completion = await request.json();
                    if (fixture.mode !== "service_proxied")
                        assert.deepEqual(completion.content, {
                            size_bytes: fixture.size_bytes,
                            checksum: claim.checksum,
                        });
                    if (fixture.fault === "completion_error") return new Response(null, { status: 503 });
                    value = {
                        ...session,
                        status: "completed",
                        completed_at_ms: 0,
                        content_ref: claim,
                        content_token: { token: "retained-token", content_ref: claim },
                    };
                } else throw new Error(`unexpected ${url}`);
                return Response.json(value);
            };
            const options = {
                baseUrl: "http://api.test",
                token: "private-token",
                headers: { "X-Private": "secret" },
                fetch: send,
            };
            const requestOptions = {
                timeoutInSeconds: fixture.fault === "timeout" ? 0.02 : 7,
                maxRetries: 3,
            };
            const result = browser
                ? new BrowserClient(options).files.prepareFileStream(
                      { namespace_alias: "demo", content: source, size_bytes: fixture.size ?? undefined },
                      requestOptions,
                  )
                : new LoonFSClient(options).files.prepareFileStream(
                      { namespace_id: "demo", content: source, size_bytes: fixture.size ?? undefined },
                      requestOptions,
                  );
            if (fixture.error) {
                await assert.rejects(result, (error) => !(error instanceof assert.AssertionError));
                assert.equal(counts.abort, fixture.fault === "completion_error" ? 0 : 1);
                assert.equal(counts.complete, fixture.fault === "completion_error" ? 1 : 0);
                if (fixture.fault === "payload_error") assert.equal(counts.payload, 1);
            } else {
                const prepared = await result;
                assert.equal(prepared.contentToken?.token, "retained-token");
                assert.equal(Buffer.concat(bodies).toString(), fixture.content);
                assert.equal(counts.complete, 1);
                assert.equal(counts.abort, 0);
            }
            assert.ok(closed, "SDK consumes or returns the source");
        });
    }

test("upload cancellation interrupts a stalled source without waiting for its cleanup", async () => {
    const scope = new TransferScope({ timeoutInSeconds: 0.02 });
    const source = new UploadSource(
        {
            async *[Symbol.asyncIterator]() {
                await new Promise(() => {});
                yield new Uint8Array();
            },
        },
        scope,
    );
    try {
        await assert.rejects(source.read());
    } finally {
        source.close();
        scope.close();
    }
});

test("upload bounds large caller chunks and validates before locking a stream", async () => {
    const scope = new TransferScope({});
    const stream = new ReadableStream<Uint8Array>();
    assert.throws(() => new UploadSource(stream, scope, -1));
    assert.equal(stream.locked, false);
    const source = new UploadSource(
        {
            async *[Symbol.asyncIterator]() {
                yield new Uint8Array(1024 * 1024);
            },
        },
        scope,
    );
    try {
        assert.equal((await source.read())?.length, 65536);
    } finally {
        source.close();
        scope.close();
    }
});

File: sdk/conformance/go/streaming_uploads_test.go
package conformance_test

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"sync/atomic"
	"testing"
	"time"

	loonfs "github.com/loonfs/loonfs-sdk-go"
	"github.com/loonfs/loonfs-sdk-go/option"
	"github.com/loonfs/loonfs-sdk-go/server"
)

type uploadCase struct {
	streamCase
	Mode  string `json:"mode"`
	Size  *int64 `json:"size"`
	Fault string `json:"fault"`
}
type uploadSource struct {
	reader   *strings.Reader
	fault    bool
	position atomic.Int64
	eof      atomic.Bool
}

func (s *uploadSource) Read(p []byte) (int, error) {
	if len(p) > 65536 {
		return 0, fmt.Errorf("unbounded source read: %d", len(p))
	}
	n, err := s.reader.Read(p)
	s.position.Add(int64(n))
	if err == io.EOF {
		s.eof.Store(true)
		if s.fault {
			err = errors.New("source failed after its last byte")
		}
	}
	return n, err
}

func TestStreamingUploads(t *testing.T) {
	data, err := os.ReadFile("../fixtures/streaming_uploads.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []uploadCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatal(err)
	}
	for _, fixture := range cases {
		t.Run(fixture.Name, func(t *testing.T) {
			source := &uploadSource{reader: strings.NewReader(fixture.Content), fault: fixture.Fault == "source_error"}
			var payload, abort, complete atomic.Int64
			var bodies bytes.Buffer // server handlers complete before their response is consumed
			claim := map[string]any{"kind": "blob", "owner_namespace_id": "demo", "content_id": "cnt_00000000000000000000000000000001", "size_bytes": fixture.SizeBytes, "checksum": map[string]any{"algorithm": fixture.Algorithm, "value": fixture.Checksum}}
			var host *httptest.Server
			host = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				path := r.URL.Path
				if path == "/object" {
					if r.Header.Get("Authorization") != "" || r.Header.Get("X-Private") != "" {
						t.Error("API credentials leaked")
					}
					if fixture.Mode == "direct_put" && r.ContentLength != *fixture.Size {
						t.Errorf("content length=%d", r.ContentLength)
					}
				} else if r.Header.Get("Authorization") != "Bearer private-token" {
					t.Error("missing API credentials")
				}
				w.Header().Set("Content-Type", "application/json")
				session := map[string]any{"namespace_id": "demo", "upload_id": "upl_test", "mode": fixture.Mode}
				access := map[string]any{"kind": "presigned_url", "method": "PUT", "url": host.URL + "/object", "expires_at_ms": 2000000000000}
				var value any
				switch {
				case strings.HasSuffix(path, "/capabilities"):
					limits := map[string]int{}
					if fixture.Mode == "direct_put" {
						limits["upload.max_content_bytes"] = 0
					}
					value = map[string]any{"protocol_version": "v0", "api_groups": []string{"filesystem/v0"}, "features": map[string]bool{"filesystem.uploads.direct_put": fixture.Mode == "direct_put", "filesystem.uploads.direct_multipart": fixture.Mode == "direct_multipart"}, "limits": limits}
				case strings.HasSuffix(path, "/uploads"):
					var body struct {
						Mode string `json:"mode"`
					}
					json.NewDecoder(r.Body).Decode(&body)
					if body.Mode != fixture.Mode {
						t.Errorf("mode %s", body.Mode)
					}
					session["checksum_algorithm"] = fixture.Algorithm
					session["part_size_bytes"] = 4
					session["access"] = access
					value = session
				case strings.HasSuffix(path, "/parts"):
					var body struct {
						Parts []struct {
							PartNumber int `json:"part_number"`
						} `json:"parts"`
					}
					json.NewDecoder(r.Body).Decode(&body)
					if len(body.Parts) != 1 {
						t.Error("must sign one part at a time")
						http.Error(w, "bad parts", 400)
						return
					}
					if source.position.Load() > payload.Load()*4+4 {
						t.Error("read the next part before sending this one")
					}
					value = map[string]any{"namespace_id": "demo", "upload_id": "upl_test", "parts": []any{map[string]any{"part_number": body.Parts[0].PartNumber, "access": access}}}
				case path == "/object" || strings.HasSuffix(path, "/content"):
					payload.Add(1)
					body, err := io.ReadAll(r.Body)
					if err != nil {
						http.Error(w, "body interrupted", 400)
						return
					}
					bodies.Write(body)
					if fixture.Fault == "timeout" {
						<-r.Context().Done()
						return
					}
					if fixture.Fault == "payload_error" {
						http.Error(w, "payload failed", 503)
						return
					}
					w.Header().Set("ETag", "test-etag")
					session["content_ref"] = claim
					value = session
				case strings.HasSuffix(path, "/abort"):
					abort.Add(1)
					session["status"] = "aborted"
					session["aborted_at_ms"] = 0
					value = session
				case strings.HasSuffix(path, "/complete"):
					complete.Add(1)
					if !source.eof.Load() {
						t.Error("completed before source EOF")
					}
					var body struct {
						Content *loonfs.UploadContentClaim `json:"content"`
					}
					json.NewDecoder(r.Body).Decode(&body)
					if fixture.Mode != "service_proxied" && (body.Content == nil || body.Content.Checksum.Value != fixture.Checksum || body.Content.SizeBytes != int64(fixture.SizeBytes)) {
						t.Error("wrong completion checksum/size")
					}
					if fixture.Fault == "completion_error" {
						http.Error(w, "completion unavailable", 503)
						return
					}
					session["status"] = "completed"
					session["completed_at_ms"] = 0
					session["content_ref"] = claim
					session["content_token"] = map[string]any{"token": "retained-token", "content_ref": claim}
					value = session
				default:
					t.Errorf("unexpected %s", path)
					http.NotFound(w, r)
					return
				}
				json.NewEncoder(w).Encode(value)
			}))
			defer host.Close()
			client := server.NewClient(option.WithBaseURL(host.URL), option.WithToken("private-token"), option.WithHTTPHeader(http.Header{"X-Private": {"secret"}}), option.WithHTTPClient(host.Client()), option.WithMaxAttempts(4))
			ctx := context.Background()
			if fixture.Fault == "timeout" {
				var cancel context.CancelFunc
				ctx, cancel = context.WithTimeout(ctx, 100*time.Millisecond)
				defer cancel()
			}
			prepared, err := client.Files.PrepareFileStream(ctx, "demo", source, fixture.Size)
			if fixture.Error {
				if err == nil {
					t.Fatal("invalid upload succeeded")
				}
				expectedAbort, expectedComplete := int64(1), int64(0)
				if fixture.Fault == "completion_error" {
					expectedAbort, expectedComplete = 0, 1
				}
				if abort.Load() != expectedAbort || complete.Load() != expectedComplete {
					t.Fatalf("abort=%d complete=%d err=%v", abort.Load(), complete.Load(), err)
				}
				if fixture.Fault == "payload_error" && payload.Load() != 1 {
					t.Error("payload retried")
				}
			} else {
				if err != nil {
					t.Fatal(err)
				}
				if prepared.ContentToken.Token != "retained-token" || bodies.String() != fixture.Content {
					t.Error("wrong prepared result or bytes")
				}
				if complete.Load() != 1 || abort.Load() != 0 {
					t.Error("unexpected session lifecycle")
				}
			}
		})
	}
}
